File: src-tauri/src/api/elaxanapi.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::time::Duration;

use crate::types::gm::Gm;

#[derive(Serialize)]
struct HandbookGi {
    search: String,
    limit: u32,
    category: Option<String>,
    language: Option<String>,
    command: Option<bool>,
    image: Option<bool>,
}

#[derive(Serialize)]
struct HandbookSr {
    search: String,
    limit: u32,
    category: Option<String>,
    language: Option<String>,
}

const BASE_URL: &str = "https://api.elaxan.com";
const ENDPOINTS: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "gi" => "/v3/gm",
    "sr" => "/v1/sr",
    "category" => "/v2/category"
};

struct ElaxanApi {
    client: Client,
}

impl ElaxanApi {
    pub fn new() -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(10))
            .build()
            .expect("Failed to create HTTP client");

        ElaxanApi { client }
    }

    pub async fn get_handbook_gi(&self, data: HandbookGi) -> Result<Vec<Gm>, reqwest::Error> {
        let endpoint = format!("{}{}", BASE_URL, ENDPOINTS["gi"]);
        let response = self.client.post(&endpoint).json(&data).send().await?;
        response.json().await
    }
}

File: src/api/elaxanApi.ts
import type { Category } from "@/types/category";
import type { APIElaXan, GmhandbookGI } from "@/types/gm";
import type { Hsr } from "@/types/hsr";
import axios from "axios";

interface HandbookGi {
    search: string[];
    limit: number;
    category?: string[];
    language?: string;
    command?: boolean;
    image?: boolean;
}

interface HandbookSr {
    search: string[];
    limit: number;
    category?: string[];
    language?: string;
}

const endpoints = {
    gi: "/v4/gm",
    sr: "/v3/sr",
    category: "/v2/category",
};

function getHandbook(
    baseURL: string,
    type: "gi",
    data: HandbookGi
): Promise<GmhandbookGI[]>;
function getHandbook(
    baseURL: string,
    type: "sr",
    data: HandbookSr
): Promise<Hsr>;
async function getHandbook(
    baseURL: string,
    type: "gi" | "sr",
    data: HandbookGi | HandbookSr
): Promise<GmhandbookGI[] | Hsr> {
    const url = new URL(endpoints[type], baseURL);

    const res = await axios.post<APIElaXan | Hsr>(url.toString(), data);
    if (type === "gi") {
        return res.data.data as GmhandbookGI[];
    }
    return res.data as Hsr;
}

async function getCategoryList(
    baseURL: string,
    type: "gi" | "sr"
): Promise<Category> {
    const url = new URL(endpoints.category, baseURL);
    const res = await axios.get(url.toString(), {
        params: {
            type,
        },
    });
    return res.data;
}

export default { getHandbook, getCategoryList };
